src/EditorFeatures/Core/Tagging/EfficientTagger.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+/// <summary>
+/// Root type of all roslyn <see cref="ITagger{T}"/> implementations.  This adds specialized hooks that allow for tags
+/// to be added to a preexisting list, rather than generating a fresh list instance.  This can help with avoiding
+/// garbage by allowing us to pool intermediary lists (especially as some high-level aggregating taggers defer to
+/// intermediary taggers).
+/// </summary>
+internal abstract class EfficientTagger<TTag> : ITagger<TTag>, IDisposable where TTag : ITag
+{
+    /// <summary>
+    /// Produce the set of tags with the requested <paramref name="spans"/>, adding those tags to <paramref name="tags"/>.
+    /// </summary>
+    public abstract void AddTags(NormalizedSnapshotSpanCollection spans, SegmentedList<ITagSpan<TTag>> tags);
+
+    public abstract void Dispose();
+
+    /// <summary>
+    /// Default impl of the core <see cref="ITagger{T}"/> interface.  Forces an allocation.
+    /// </summary>
+    public IEnumerable<ITagSpan<TTag>> GetTags(NormalizedSnapshotSpanCollection spans)
```

Comment:
> IEnumerable<ITagSpan<TTag>> [](http://example.com/codeflow?start=10&length=28)

I'm going to ask a colossally dumb question here, but if we made the *massive* assumption that platform only iterated over the result we passed back once, couldn't we implement an IEnumerator whose dispose returned this to the cache?

From what I see here ([https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS-Platform?path=/src/Editor/Text/Impl/TagAggregator/TagAggregator.cs&version=GBmain&line=487&lineEnd=488&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS-Platform?path=/src/Editor/Text/Impl/TagAggregator/TagAggregator.cs&version=GBmain&line=487&lineEnd=488&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) ), it looks like that might be the case.

It does look like they are properly disposing the enumerator.

Reply:
not a dumb question.  

but it's really much too much of an assumption IMO.  These are apis that *anyone* can call.  And it would be very very very very unsafe to do this.

Note: IEnumerators are disposable, not IEnumerables.  An IEnumerable is expected to be able to eb called numerous times without issues.  

even if VS was ok now, they could trivially change and blow up.

Reply:
Yeah, I know it would make the enumerable we pass back only work one time, but the appeal of no allocations was too much for me to not mention the idea.